File: backend/app/core/upgrades/effects/armor.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Armor(UpgradeImplementation):
    id = 'armor'
    name = 'Hardened Shell'
    description = 'Reduce incoming damage by 2.5%'
    rarity = 'rare'
    type = 'effect'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.effects.register_effect('protection', {'amount': 0.025})

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/effects/armor.ts
import { Upgrade, type UpgradeDef, type DamageRef } from '../Upgrade'
import { RarityID, UpgradeTypeID } from '../../data/ID'

export const ArmorDef: UpgradeDef = {
  id: "armor",
  name: "Hardened Shell",
  description: "Reduce incoming damage by 2.5%",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.Effect,
  cost: 10,
  effect: "protection",
  effectValue: 0.025,
  stackable: true,
  maxStacks: 7,
}

export class Armor extends Upgrade {
  onApply(): void {}

  modifyPlayerHurt(damage: DamageRef): void {
    // Always let at least 1 damage through
    damage.amount = Math.max(1, damage.amount * (1 - this.def.effectValue!))
  }
}
